### DISCUSSION TITLE
OG image generation

### DISCUSSION THREAD

**Discussion by taishi55** [sveltejs/kit#8299] [Q&A] [answered]
### Describe the problem

Hello, I'm looking for providing easy way to dynamically generate open-graph images using Sveltekit.

This function exists in Next.js but not in Sveltekit. 
https://www.npmjs.com/package/next-api-og-image

### Describe the proposed solution

Is it possible to make something like https://www.npmjs.com/package/next-api-og-image?

### Importance

nice to have

**Comment by benmccann** [comment]
See https://geoffrich.net/posts/svelte-social-image/ and https://github.com/sveltejs/sites/pull/412 for examples

**Reply by taishi55** [reply]
Thank you! The integration of Vercel plugin seems perfect solution for me. Should I install `satori ` by `npm i satori  -D`? or is it installed in Sveltekit by default?

**Reply by benmccann** [reply]
It is not installed by SvelteKit by default, so you will need to install it

**Comment by archiewood** [comment]
Does anyone have a good pattern for static image generation?

I.e. using adapter-static and generating a templated OG image for each page, based on, for example, the title and description.

I've been playing around with some options but unclear where to store  files to make it DRY:

## single SEO route

/og.png?title=Foo&description=bar

problem: cannot access the search params in SSG

## many SEO routes

/foo/og.png

problem: needing to create many +server.js files is a pain

something else?

**Reply by david-plugge** [reply]
What about a `[...path]/og.png/+server.js` file which handles all your Images?

**Reply by archiewood** [reply]
Good idea, 
How do I access the title and description data for each of the routes?

EDIT:
I was able to create an API route which exposed this information, and fetch it.

**Reply by benmccann** [reply]
You can also do `/og/[title]/[description]/image.png`

**Reply by archiewood** [reply]
@benmccann this was an interesting approach i experimented with, but it didn't handle cases where the title or description were missing, which is the case for many of our pages

For anyone interested in the final approach i went for, see this PR: https://github.com/evidence-dev/evidence/pull/2985
- `src/lib/SocialCard.svelte`
  - the layout of the social card
- `src/routes/[...route]/og.png/+server.js`
  - generate the images, fetch the title and description from an API endpoint and filter for the relevant page
- `src/routes/+layout.svelte`
  - add the og links for each page in the `<head>` tags
- `src/routes/api/pagesManifest.json/+server.js`
   - the API endpoint that exposes the title and description of every page

**Comment by theetherGit** [comment, marked as answer]
For future reference, take a look at [sveltekit-og](https://sveltekit-og.dev/docs): 
1. [Pre-render `og.png` for docs](https://sveltekit-og.dev/docs/examples/pre-render)
2. [Auto generate `og.png` routes](https://sveltekit-og.dev/docs/examples/auto-generate-routes)
